internal/hud/server/websocket_reader.go

```diff
@@ -0,0 +1,174 @@
+		handler: NewLogStreamer(),
+	}
+}
+
+type ViewHandler interface {
+	Handle(v proto_webview.View) error
+}
+
+type LogStreamer struct {
+	logstore   *logstore.LogStore
+	printer    *hud.IncrementalPrinter
+	checkpoint logstore.Checkpoint
+}
+
+func NewLogStreamer() *LogStreamer {
+	// TODO(maia): wire this
+	printer := hud.NewIncrementalPrinter(hud.Stdout(colorable.NewColorableStdout()))
+	return &LogStreamer{
+		logstore: logstore.NewLogStore(),
+		printer:  printer,
+	}
+}
+func (ls *LogStreamer) Handle(v proto_webview.View) error {
+	fromCheckpoint := logstore.Checkpoint(v.LogList.FromCheckpoint)
+	toCheckpoint := logstore.Checkpoint(v.LogList.ToCheckpoint)
+
+	if fromCheckpoint == -1 {
+		// Server has no new logs to send
+		return nil
+	}
+
+	segments := v.LogList.Segments
+	if fromCheckpoint < ls.checkpoint {
+		// The server is re-sending some logs we already have, so slice them off.
+		deleteCount := ls.checkpoint - fromCheckpoint
+		segments = segments[deleteCount:]
+	}
+
+	// TODO(maia): filter for the resources that we care about (`tilt logs resourceA resourceC`)
+	//   --> and if there's only one resource, don't prefix logs with resource name?
+	for _, seg := range segments {
+		// TODO(maia): secrets???
+		ls.logstore.Append(webview.LogSegmentToEvent(seg, v.LogList.Spans), model.SecretSet{})
+	}
+
+	ls.printer.Print(ls.logstore.ContinuingLines(ls.checkpoint))
+
+	if toCheckpoint > ls.checkpoint {
+		ls.checkpoint = toCheckpoint
+	}
+
+	return nil
+}
+
+func (wsr *WebsocketReader) Listen() {
+	// catch signals to kill
+	// TODO: use signal catching we already use in `up` etc.
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
+
+	log.Printf("connecting to %s", wsr.url.String())
+
+	c, _, err := websocket.DefaultDialer.Dial(wsr.url.String(), nil)
+	if err != nil {
+		log.Fatal("dial:", err)
+	}
+	defer c.Close()
+
+	done := make(chan struct{})
+
+	go func() {
+		// TODO(maia): make sure this closes okay 😅
+		defer close(done)
+		for {
+			_, data, err := c.ReadMessage()
+			if err != nil {
+				log.Println("🚨 error reading:", err)
+				return
+			}
+
+			v := proto_webview.View{}
+			unmarshaller := jsonpb.Unmarshaler{}
+			err = unmarshaller.Unmarshal(bytes.NewReader(data), &v)
+			if err != nil {
+				log.Println("🚨 error unmarshalling:", err)
+			}
+			err = wsr.handler.Handle(v)
+			if err != nil {
+				log.Println("🚨 handler error:", err)
+			}
+
+			toCheckpoint := v.LogList.ToCheckpoint
+			if toCheckpoint > 0 {
+				// If server is using the incremental logs protocol, ack the
+				// message so the next time the websocket sends data, it only
+				// sends logs from here on forward
+				resp := proto_webview.AckWebsocketRequest{
+					ToCheckpoint:  toCheckpoint,
+					TiltStartTime: v.TiltStartTime,
+				}
+				marshaller := jsonpb.Marshaler{OrigName: false, EmitDefaults: true}
+				respJson, err := marshaller.MarshalToString(&resp)
```

Comment:
idiomatically you'd just plug the io streams together like
writer := c.NextWriter(TextMessage)
defer writer.Close()
marshaller.Marshal(writer, msg)

to avoid converting back and forth from bytes to strings (which puts a lot of stress on the garbage collector)